File: spaceinvaders.py
from tkinter import *
from array import *
import time
import math
import random
from playsound import playsound
from threading import Thread
from PIL import Image,ImageTk

canvHeight = 600
canvWidth = 1280
t=Tk()
c = Canvas(t, height=canvHeight, width=canvWidth)
c.pack()
c.create_rectangle(0, 0, canvWidth, canvHeight, fill="black")
listOfAliens = []
listOfAliensImg = []
angles = []
outerLoop = []
innerLoop = []
alienColors = ["white", "lightgreen", "red", "yellow", "skyblue"]
shotList = []
numParticles = 0

alien_1_proper = Image.open(".\\aliens\\alien1.png")
alien_1 = PhotoImage(file=".\\aliens\\alien1.png")
alien_2_proper = Image.open(".\\aliens\\alien2.png")
alien_2 = PhotoImage(file=".\\aliens\\alien2.png")

alien_1_rotated = ImageTk.PhotoImage(alien_1_proper.rotate(40))

#ship_main = c.create_oval((canvWidth/2)-40, 510, (canvWidth/2)+40, 590, outline="skyblue", fill="skyblue")
ship_inner = c.create_polygon((canvWidth/2), 510, (canvWidth/2)+32, 580, (canvWidth/2)-32, 580,outline="skyblue", fill="#db655c")

def moveShip(event):
    x, y = event.x, event.y
    c.moveto(ship_inner, x-32, y-50)

def cloneAliens(pattern, startX, startY, waves, lengthOfWave, offset):
    global numParticles
    for i in range(lengthOfWave):
        numParticles += 1
        #alien = c.create_rectangle((startX-15)-(i*offset), startY-15, (startX+15)-(i*offset), startY+15, fill=alienColors[int(pattern[i%len(pattern):i%len(pattern)+1])])#alienColors[pattern[i % len(pattern)]])
        alien = c.create_image((startX-20)-(i*offset), startY-20, image=alien_1)
        listOfAliens.append(alien)
        listOfAliensImg.append(alien_1_proper)
        angles.append(0)
        outerLoop.append(0)
        innerLoop.append(0)

    print(listOfAliensImg)
commandList = "-"

def cloneShots():
    global numParticles
    numParticles+=1
    coords = c.coords(ship_inner)
    #for i in range(10):
    shots = c.create_rectangle(coords[0]-4, coords[1]-40, coords[0]+4, coords[1], fill="lightblue")
    shotList.append(shots)

def moveShots(speed):
    global numParticles
    i = 0
    while i < len(shotList):
        c.move(shotList[i], 0, 0-speed)
        i+=1
        '''print(i)
        if(c.coords(shotList[i])[1] < 0):
            shotList.pop(i)
            c.delete(shotList[i])
            numParticles -= 1
            i -=1
        i+=1'''


centralX, centralY, centralShotX, centralShotY = 0, 0, 0, 0
def soundEffect():
    playsound("C:\\Users\\mchlc\\mystuff\\spaceinvaders\\gun-gunshot-01.wav")

def deathAnimation(count, xPos, yPos):
    pass

def moveAliens(command, speed):
    commandList = command.split('-')
    
    #loop.append(0)
    i = 0
    a=ImageTk.PhotoImage(alien_2_proper)
    while i < len(listOfAliens):
        cmd = commandList[outerLoop[i]]
        for char in cmd:
            if char.isalpha():
                breakPoint = cmd.index(char)
                breakLetter = char
                break

        if(c.coords(listOfAliens[i])[0] + 15 > 0):
            if (breakLetter == "R"):
                if (innerLoop[i] < int(cmd[0:breakPoint])):
                    innerLoop[i] += 1
                  
                    c.itemconfig(listOfAliens[i], image=alien_1_rotated)
                    angles[i]+= int(cmd[breakPoint+1:len(cmd)])/360*math.pi*2
                    c.move(listOfAliens[i], speed*math.cos(angles[i]), speed*math.sin(angles[i]))
                else:#need to add terminating condition
                    if (outerLoop[i] < len(commandList)-1):
                        innerLoop[i] = 0
                        outerLoop[i] += 1
                    
            elif (breakLetter == "L"):
                if (innerLoop[i] < int(cmd[0:breakPoint])):
                    innerLoop[i] += 1
                    angles[i] -= int(cmd[breakPoint+1:len(cmd)])/360*math.pi*2
                    c.move(listOfAliens[i], speed*math.cos(angles[i]), speed*math.sin(angles[i]))
                else:#need to add terminating condition
                    if (outerLoop[i] < len(commandList)-1):
                        innerLoop[i] = 0
                        outerLoop[i] += 1

            else:
                if (innerLoop[i] < int(cmd[0:breakPoint])):
                    innerLoop[i] += 1
                    c.move(listOfAliens[i], int(cmd[breakPoint+1:len(cmd)])*math.cos(angles[i]), int(cmd[breakPoint+1:len(cmd)])*math.sin(angles[i]))
                else:#need to add terminating condition
                    if (outerLoop[i] < len(commandList)-1):
                        innerLoop[i] = 0
                        outerLoop[i] += 1
            
            k = 0
            while k < len(shotList): #detect collision
                #alien x and y positions
                centralX = c.coords(listOfAliens[i])[0]+15 
                centralY = c.coords(listOfAliens[i])[1]+15
                #shot x and y positions
                centralShotX = c.coords(shotList[k])[0]+3
                centralShotY = c.coords(shotList[k])[1]+20
                if ((centralShotX-3 < centralX+15) and (centralShotX+3> centralX-15) and
                    (centralShotY-20 < centralY+15) and (centralShotY+20 > centralY-15)):
                    #delete alien
                    c.create_text(centralX, centralY, text="10", fill="white", font=("Segoe UI Bold", 20))
                    c.delete(listOfAliens[i])
                    listOfAliens.pop(i)
                    outerLoop.pop(i)
                    innerLoop.pop(i)
                    angles.pop(i)
                    i -= 1
                    
                    global numParticles
                    numParticles-=1

                    #t = Thread(target=soundEffect)
                    #t.start()

                    c.delete(shotList[k])
                    shotList.pop(k)
                    k -= 1
                k+= 1
                
        else:
            c.move(listOfAliens[i], speed, 0)
        
        i+=1
alienSpeed = 6
cloneAliens("01234", 0, 350, 1, 25, alienSpeed*15) #appears every 10 frames

c.bind('<Motion>', moveShip)

startTime = time.time()
print(startTime)
while True: #main game loop
    newTime = time.time()
    if(newTime - startTime > 0.2):
        cloneShots()
        startTime = newTime
    moveShots(11)
    moveAliens("50F6-30L3-35R3-50F6-120L3-65F6", alienSpeed)
    t.update()
    if(numParticles < 60):
        time.sleep(0.001)
    
    
    
    

File: spaceinvaders.js
const canvas = document.getElementById("myCanvas");
const c = canvas.getContext("2d");
c.imageSmoothingEnabled = false;
collisionRadius = 15;
borderThickness = collisionRadius/5
alienSpeed = 3
shipSize = 35

canvas.setAttribute('height', window.innerHeight + "px");
canvas.setAttribute('width', window.innerWidth + "px");

var listOfAliens = []
var memoryCommandList = []
var shots = []
var deathParticles = []
var xPos = -1
var yPos = -1
var waveInfo = {wavelength: -1, commandList: [], pattern: -1, currentID: -1, startX: -1, startY: -1, initDirection: -1}

//sounds
var shotSound = new Audio("./sounds/laser.wav")


function cloneAlien(startX, startY, type, angle){
    //console.log(type + angle)
    //console.log(startX + startY)
 
    c.save()
    c.beginPath()
    c.translate(startX, startY)
    c.rotate(angle)
    //c.rect(startX-collisionRadius, startY-collisionRadius, collisionRadius*2, collisionRadius*2)
    c.rect(0-collisionRadius, 0-collisionRadius, collisionRadius*2, collisionRadius*2)
    c.fillStyle = "gray";
    c.fill();
    c.closePath();

    c.beginPath()
    c.rect(0-collisionRadius+borderThickness, 0-collisionRadius+borderThickness, (collisionRadius*2)-(borderThickness*2), (collisionRadius*2)-(borderThickness*2))
    c.fillStyle = "#00ff08";
    c.fill()
    c.closePath()

    c.lineWidth = borderThickness+2
    c.beginPath()
    c.moveTo(0-borderThickness, 0-collisionRadius+borderThickness)
    //c.lineTo(startX+(Math.sqrt(3)-1.5)*collisionRadius+borderThickness, startY)
    c.lineTo(0+collisionRadius-borderThickness, 0)
    c.lineTo(0-borderThickness, 0+collisionRadius-borderThickness)
    c.lineTo(0-borderThickness, 0-collisionRadius+borderThickness)
    c.strokeStyle = "gray"
    c.stroke()
    c.fillStyle = "red"
    c.fill()

    c.restore()
}

function initWave(command, wavelength, pattern, startX, startY, initDirection){
    waveInfo.commandList = command.split('-')
    waveInfo.wavelength = wavelength
    waveInfo.pattern = pattern
    waveInfo.currentID = 0
    waveInfo.startX = startX
    waveInfo.startY = startY
    waveInfo.initDirection = initDirection
    for (let i = 0; i < waveInfo.commandList.length; i++){
        let subcmd = waveInfo.commandList[i]
        for (let j = 0; j < subcmd.length; j++){
            if(isAlpha(subcmd.charAt(j))){
                memoryCommandList.push({subCommand: subcmd, breakpoint: j, breakletter: subcmd.charAt(j)})
                break
            }
        }
    }
}

function initAlien(){
    if (waveInfo.currentID < waveInfo.wavelength){
        listOfAliens.push({x: waveInfo.startX, y: waveInfo.startY, angle: waveInfo.initDirection, outerLoop: 0, innerLoop: 0, /*species: pattern.charAt(i%pattern.length),*/ wave: "original"})
        waveInfo.currentID +=1;
    }
}

function cloneShots(){
    let i = 0;
    while (i<shots.length){
        c.beginPath()
        c.rect(shots[i].x-3, shots[i].y-30, 6, 40)
        c.fillStyle = "lightblue"
        c.fill()
        c.closePath()
        i++
    }
    
}

function initShot(x, y, type){
    shots.push({x: x, y: y, type: type})
}

function updateShotCoords(){
    let i = 0;
    while (i < shots.length){
        if (shots[i].type == "normal"){
            shots[i].y -=7
        }else if (shots[i].type == "fast"){
            shots[i].y -= 15
        }

        if (shots[i].y <= 0){
            shots.splice(i, 1)
            i--
        }
        i++
    }
}

function drawShip(x, y){
    const img = new Image()
    img.src = "./aliens/ship1.png"
    c.drawImage(img, x-shipSize, y-shipSize, shipSize*2, shipSize*2)
}

function isAlpha(ch){
    return /^[A-Z]$/i.test(ch);
}

function updateAlienCoords(){
    let i = 0;
    while (i < listOfAliens.length){
        let outerLoopLocation = memoryCommandList[listOfAliens[i].outerLoop]
        let angle;
        if(listOfAliens[i].outerLoop < waveInfo.commandList.length){
            if(outerLoopLocation.breakletter == "R"){
                if(listOfAliens[i].innerLoop < parseInt(outerLoopLocation.subCommand.substring(0, outerLoopLocation.breakpoint))){
                    angle = parseFloat(outerLoopLocation.subCommand.substring(outerLoopLocation.breakpoint+1))
                    listOfAliens[i].angle += (angle/360)*(2*Math.PI)
                    listOfAliens[i].innerLoop += 1;
                    listOfAliens[i].x += alienSpeed*Math.cos(listOfAliens[i].angle)
                    listOfAliens[i].y += alienSpeed*Math.sin(listOfAliens[i].angle)
                }else{
                    listOfAliens[i].innerLoop = 0;
                    listOfAliens[i].outerLoop += 1;
                }
    
            }else if (outerLoopLocation.breakletter == "L"){
                if(listOfAliens[i].innerLoop < parseInt(outerLoopLocation.subCommand.substring(0, outerLoopLocation.breakpoint))){
                    angle = parseFloat(outerLoopLocation.subCommand.substring(outerLoopLocation.breakpoint+1))
                    listOfAliens[i].angle -= (angle/360)*(2*Math.PI)
                    listOfAliens[i].innerLoop += 1;
                    
                    listOfAliens[i].x += alienSpeed*Math.cos(listOfAliens[i].angle)
                    listOfAliens[i].y += alienSpeed*Math.sin(listOfAliens[i].angle)
                }else{
                    listOfAliens[i].innerLoop = 0;
                    listOfAliens[i].outerLoop += 1;
                }
    
            }else{
                if(listOfAliens[i].innerLoop < parseInt(outerLoopLocation.subCommand.substring(0, outerLoopLocation.breakpoint))){
                    listOfAliens[i].innerLoop += 1;
                    listOfAliens[i].x += 3*Math.cos(listOfAliens[i].angle)
                    listOfAliens[i].y += 3*Math.sin(listOfAliens[i].angle)
                }else{
                    listOfAliens[i].innerLoop = 0;
                    listOfAliens[i].outerLoop += 1;
                }
            }

            //detect collision
            for (let j = 0; j < shots.length; j++){
                if ((shots[j].x-3 < listOfAliens[i].x+collisionRadius) && 
                (shots[j].x+3> listOfAliens[i].x-collisionRadius) &&
                (shots[j].y-20 < listOfAliens[i].y+collisionRadius) && 
                (shots[j].y+20 > listOfAliens[i].y-collisionRadius)){
                    let originX = listOfAliens[i].x
                    let originY = listOfAliens[i].y
                    listOfAliens.splice(i, 1)
                    i--;
                    shots.splice(j, 1)
                    j++;

                    //initiate particles
                    let subDeathParticles = []
                    let radius = 5
                    let numParticles = Math.floor(Math.random()*10)+5
                    let particleColors = ["gray", "#00ff08", "red"]
                    for (let k = 0; k < numParticles; k++){
                        subDeathParticles.push({
                            x: originX+(radius*Math.cos(k*((Math.PI*2)/numParticles))),
                            y: originY+(radius*Math.sin(k*((Math.PI*2)/numParticles))),
                            variationX: (Math.random())-0.5,
                            variationY: (Math.random()*1)-1,
                            angle: k*((Math.PI*2)/numParticles),
                            size: Math.floor(Math.random()*4)+1,
                            initGravity: 3,
                            particleColor: particleColors[Math.floor(Math.random()*3)],
                            loop: 0
                        })
                    }
                    deathParticles.push(subDeathParticles)
                }
            }
        }else{
            if(listOfAliens[i].outerLoop >= waveInfo.commandList.length){
                listOfAliens.splice(i, 1)
            }
        }
        i++;
    }
}

function cloneAliens(){
    for (let i = 0; i < listOfAliens.length; i++){
        cloneAlien(listOfAliens[i].x, listOfAliens[i].y, 1, listOfAliens[i].angle)
    }
}

/*function hexToRgb(hex) {
    var result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
    return result ? {
      r: parseInt(result[1], 16),
      g: parseInt(result[2], 16),
      b: parseInt(result[3], 16)
    } : null;
  }*/
  
function updateDeathParticles(){
    let i = 0;
    let j = 0;
    while (i < deathParticles.length){
        j = 0;
        while (j < deathParticles[i].length){
            if(deathParticles[i][j].y < 625){
                deathParticles[i][j].x += Math.cos(deathParticles[i][j].angle)*0.5 + deathParticles[i][j].variationX
                deathParticles[i][j].y -= Math.sin(deathParticles[i][j].angle)*0.5 + deathParticles[i][j].initGravity + deathParticles[i][j].variationY 
                deathParticles[i][j].initGravity -= 0.1
                deathParticles[i][j].loop++;
                j++
            }else{
                deathParticles[i].splice(j, 1)
            }
        }
        if(deathParticles[i].length == 0){
            deathParticles.splice(i, 1)
            i--;
        }
        i++;
    }
}

let size;
function drawDeathParticles(){
    for(let i = 0; i < deathParticles.length; i++){
        for(let j = 0; j < deathParticles[i].length; j++){
            size = deathParticles[i][j].size
            c.beginPath()
            c.rect(deathParticles[i][j].x-size, deathParticles[i][j].y-size, size*2, size*2)
            c.fillStyle = deathParticles[i][j].particleColor;
            c.fill()
            c.closePath()
        }
    }
}

initWave("120F6-45L2-35R3-120F6-180L1.5-100F6-20R7-100F6-45L2-150F6", 45, "1254", 0, 300, 0)

let start = Date.now();


function frame(event){
    c.clearRect(0, 0, canvas.width, canvas.height);
    cloneAliens()
    updateAlienCoords()
    //console.log(listOfAliens)
    drawShip(xPos, yPos-5)
    if(Date.now() - start > 120){
        initShot(xPos, yPos-10, "fast")
        start = Date.now()
    }
    updateShotCoords()
    cloneShots()
    updateDeathParticles()
    drawDeathParticles()
}

function updateMousePos(event){
    let rect = canvas.getBoundingClientRect();
    xPos = event.clientX - rect.left;
    yPos = event.clientY - rect.top;
}

document.getElementById("myCanvas").addEventListener("mousemove", updateMousePos)
var interval = setInterval(frame, 10)
var alienInterval = setInterval(initAlien, 250)



